slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbSlashingProtection.java

```diff
@@ -108,104 +107,38 @@ public boolean maySignAttestation(
-        LOG.warn(
-            "Detected double signed attestation {} for {}", existingAttestation.get(), publicKey);
-        return false;
-      } else {
-        // same slot and signing_root is allowed for broadcasting previous attestation
-        return true;
-      }
-    }
-
-    final Optional<UInt64> minimumSourceEpoch =
-        signedAttestationsDao.minimumSourceEpoch(h, validatorId);
-    if (minimumSourceEpoch.map(minEpoch -> sourceEpoch.compareTo(minEpoch) < 0).orElse(false)) {
-      LOG.warn(
-          "Attestation source epoch {} is below minimum existing attestation source epoch {}",
-          sourceEpoch,
-          minimumSourceEpoch.get());
-      return false;
-    }
-
-    final Optional<UInt64> minimumTargetEpoch =
-        signedAttestationsDao.minimumTargetEpoch(h, validatorId);
-    if (minimumTargetEpoch.map(minEpoch -> targetEpoch.compareTo(minEpoch) <= 0).orElse(false)) {
-      LOG.warn(
-          "Attestation target epoch {} is below minimum existing attestation target epoch {}",
-          targetEpoch,
-          minimumTargetEpoch.get());
-      return false;
-    }
-
-    // check that no previous vote is surrounding the attestation
-    final Optional<SignedAttestation> surroundingAttestation =
-        signedAttestationsDao.findSurroundingAttestation(h, validatorId, sourceEpoch, targetEpoch);
-    if (surroundingAttestation.isPresent()) {
-      LOG.warn(
-          "Detected surrounding attestation {} for attestation signingRoot={} sourceEpoch={} targetEpoch={} publicKey={}",
-          surroundingAttestation.get(),
-          signingRoot,
-          sourceEpoch,
-          targetEpoch,
-          publicKey);
-      return false;
-    }
-
-    // check that no previous vote is surrounded by attestation
-    final Optional<SignedAttestation> surroundedAttestation =
-        signedAttestationsDao.findSurroundedAttestation(h, validatorId, sourceEpoch, targetEpoch);
-    if (surroundedAttestation.isPresent()) {
-      LOG.warn(
-          "Detected surrounded attestation {} for attestation signingRoot={} sourceEpoch={} targetEpoch={} publicKey={}",
-          surroundedAttestation.get(),
-          signingRoot,
-          sourceEpoch,
-          targetEpoch,
-          publicKey);
-      return false;
-    }
+          lockForValidator(handle, LockType.ATTESTATION, validatorId);
 
-    final SignedAttestation signedAttestation =
-        new SignedAttestation(validatorId, sourceEpoch, targetEpoch, signingRoot);
-    signedAttestationsDao.insertAttestation(h, signedAttestation);
-    return true;
+          if (attestationValidator.directlyConflictsWithExistingEntry()
+              || attestationValidator.isSurroundedByExistingAttestation()
+              || attestationValidator.surroundsExistingAttestation()) {
+            return false;
+          } else if (attestationValidator.existsInDatabase()) {
+            return true;
+          } else if (attestationValidator.hasSourceOlderThanWatermark()
```

Comment:
Should the watermark checks come last? Before they were being done between the existing & conflict checks and the surround and surrounding checks.

Reply:
Yeah, they need to, as, even if a low-watermark exists, if I match an existing then I can sign it.
Effectively, in the past we did the "do I exist" as a precheck - and didn't care if others also existed.
What we've really done, is separate the concept of "do others exist" and "do I exist".